### ISSUE TITLE
Network play

### ISSUE THREAD

**Issue by ymartel06** [mupen64plus/mupen64plus-core#112]
Hi,

To begin, thanks everyone for your great work on Mupen64plus, it's awesome.

I'm working to make the netplay for mupen64plus and I posted a message in google code few days ago.
https://code.google.com/p/mupen64plus/issues/detail?id=16&colspec=ID%20Type%20Component%20Status%20Priority%20Stars%20Milestone%20Owner%20Summary
It's my bad, you moved on github :)

I'm aiming to have the netplay for maximum 4 network players with all combinations available. (Multiple locals vs network etc...)

This morning I pushed a version with the support of 2 players (One by computer)
https://github.com/ymartel06/mupen64plus-core/commit/6fe81b909bf13dd4abbf611b20d615748290001d

To test it, get my console https://github.com/ymartel06/mupen64plus-ui-console and my core https://github.com/ymartel06/mupen64plus-core/.
Then you can launch the server with:
mupen64plus-ui-console.exe --server your-rom
For the client:
mupen64plus-ui-console.exe --client IP your-rom

Currently, it's only TCP (UDP will come really later) and the default is 6464 (you can change it using --server-port).

Before to work on the 4 players, I will work on:
* Fix the clock to take care of CLOCKS_PER_SEC for the client. I use a clock sync to be able to launch the emulator at around the same time.
* Don't allow to press pause, forward, backyard and more, when the emulator is in network
* Try to build it on Linux / MacOS and fix what is needed (I try to make everything cross platform but I can have a bad surprise)
* Add an option to force the frame sync between players (good to have the same simulation but can be laggy, looks like this concept http://www.gamasutra.com/view/feature/131503/1500_archers_on_a_288_network_.php)

Known issue:
* When you move the render window, you can have some disconnection.

Of course feedbacks are welcome. If you want some explanations of what I made (like flow or technical), I will try my best to answer.

Note: It seems I can't open a discussion.

**Comment by tony971** [comment]
This could be good with 64DD emulation if mupen ever adds that functionality. I know that cen64 and MESS have made huge leaps in it.

**Comment by wareya** [comment]
Time travel / rollback when?

(I jest. Even desync-free low lag emu netplay is an accomplishment at this point!)

**Comment by Narann** [comment]
Hello @ymartel06! Nice to see you here! I encourage peoples to test and feedback on this precious development netplay is.

**Comment by richard42** [comment]
Yeah this is great to see.  Users have been wanting netplay for a very long time.

**Comment by Biofreak1987** [comment]
It would be nice to see following features once you're done with 4 player netplay and udp support:

* Improved golf mode ([see aqz-netplay for more info](https://www.play64.com/netplay-plugin/)):
Make it an extra key instead of using a controller button, would be handy for taking turns in single player games.
* Sync save files before the game starts:
[an early version of dolphin dualcore-netplay](https://dolphin-emu.org/download/list/dc-netplay/1/) used to temp. sync memcard files, which made it much easier to start a netplay session.

... but even without these features it would be already a huge improvement over the current netplay situation for n64 emulators :)

**Comment by ymartel06** [comment]
In my case, the network isn't frame sync by default. So the golf mode isn't needed.
About sharing the memcard files but also maybe the rom, it's an excellent idea but I will work on it when everything will work correctly.
I have pushed some changes this morning and I will keep continue to work on it.

**Comment by ymartel06** [comment]
Hi,

It's now possible to play with more than one local player by computer.

To test this functionality, I added a new parameter: local-player.
If it's not set, the default is 1.

So, if you want to do a 2 vs 2, the configuration should like:
mupen64plus-ui-console.exe --server --local-player=2 your-rom
mupen64plus-ui-console.exe --client IP --local-player=2 your-rom

All combinations work like 1 vs 1, 2 vs 1, 3 vs 1... you must have at least 2 players and the maximum is 4.

Another great news, Linux build works (at least compile) now, if someone wants to test and gives feedback, it will be appreciated. The MacOs version should also compile.
I compiled it on Ubuntu 14.04 with g++ without any issues.


I'm currently trying to fix on huge on issue on the windows version.
When we drag or resize the render window, the emulation "pauses".
This issue introduces delay and disconnection inside my network stuff.
I tried to use SDL 2.0 (really easy to integrate it) and I've got the exact same behavior.
It's coming from windows, and I can't find any good solution to fix it.
I read several topics about it and I can't find any good idea.
http://www.gamedev.net/topic/488074-win32-message-pump-and-opengl---rendering-pauses-while-draggingresizing/
http://www.gamedev.net/community/forums/topic.asp?topic_id=440341
http://www.tech-archive.net/Archive/Development/microsoft.public.win32.programmer.ui/2006-02/msg00153.html

If someone wants to help, it's welcome!

Cheers,

Yohann

**Comment by Narann** [comment]
Great news!

> This issue introduces delay and disconnection inside my network stuff.

Maybe a silly idea but wouldn't be a way for the _resizer_ to send a _user_resizing_ to server. This way server hang other sessions (_x_ amount of time max) then _resizer_ send _user_resize_done_ to the server when resize is over.

**Comment by wareya** [comment]
you need to change the way the video window is created to get rid of the stall.

**Comment by atcasanova** [comment]
$ ./mupen64plus --verbose --server /home/klingon/Mario\ Kart\ 64\ \(USA\).n64
(...)
UI-Console: attached to core library 'Mupen64Plus Core' version 2.5.0
UI-Console:             Includes support for Dynamic Recompiler.
UI-Console:             Includes support for MIPS r4300 Debugger.
Segmentation fault
$ echo $?
139

**Comment by Mastergatto** [comment]
I just stumbled across this open source project: https://github.com/pond3r/ggpo

Do you think it could be used and integrated with mupen64plus?
